contracts/JBV3Token.sol

```diff
@@ -0,0 +1,355 @@
+  ) external override {
+    // Can't transfer for a wrong project.
+    if (projectId != 0 && _projectId != projectId) revert BAD_PROJECT();
+
+    transfer(_to, _amount);
+  }
+
+  /** 
+    @notice
+    Transfer tokens between accounts.
+
+    @param _projectId The ID of the project to which the token belongs. This is ignored.
+    @param _from The originating address.
+    @param _to The destination address.
+    @param _amount The amount of the transfer, as a fixed point number with 18 decimals.
+  */
+  function transferFrom(
+    uint256 _projectId,
+    address _from,
+    address _to,
+    uint256 _amount
+  ) external override {
+    // Can't transfer for a wrong project.
+    if (projectId != 0 && _projectId != projectId) revert BAD_PROJECT();
+
+    transferFrom(_from, _to, _amount);
+  }
+
+  function transferOwnership(uint256 _projectId, address _newOwner) external override {
+    _projectId; // Prevents unused var compiler and natspec complaints.
+
+    return super.transferOwnership(_newOwner);
+  }
+
+  /** 
+    @notice
+    Migrate v1 & v2 tokens to v3.
+  */
+  function migrate() external {
+    uint256 _fundsToMigrateFromV1;
+    uint256 _fundsToMigrateFromV2;
+    if (address(v1TicketBooth) != address(0)) {
+      // Get a reference to the v1 projects ERC20 tokens.
+      ITickets _v1Token = v1TicketBooth.ticketsOf(projectId);
+      _fundsToMigrateFromV1 = _migrateV1Tokens(_v1Token);
+    }
+
+    if (address(v2TokenStore) != address(0)) {
+      // Get a reference to the v2 project's ERC20 tokens.
+      IJBToken _v2Token = v2TokenStore.tokenOf(projectId);
+      _fundsToMigrateFromV2 = _migrateV2Tokens(_v2Token);
+    }
+
+    uint256 _tokensToMint = _fundsToMigrateFromV1 + _fundsToMigrateFromV2;
+    // mint tokens directly
+    _mint(msg.sender, _tokensToMint);
+  }
+
+  /** 
+    @notice
+    Migrate v1 tokens to v3.
+    @param _v1Token The v1 token instance.
+
+    @return amount of v2 tokens to be migrated
+  */
+  function _migrateV1Tokens(ITickets _v1Token) internal returns(uint256) {     
+    // Get a reference to the migrator's unclaimed balance.
+    uint256 _tokensToMintFromUnclaimedBalance = v1TicketBooth.stakedBalanceOf(msg.sender, projectId);
+
+    // Get a reference to the migrator's ERC20 balance.
+    uint256 _tokensToMintFromERC20s = _v1Token == ITickets(address(0)) ? 0 : _v1Token.balanceOf(msg.sender);
+    
+    // total v3 tokens to mint
+    uint256 v3TokensToMint = _tokensToMintFromERC20s + _tokensToMintFromUnclaimedBalance;
+
+    if (v3TokensToMint == 0)
+      return 0;
+
+    // Transfer v1 ERC20 tokens to this contract from the msg sender if needed.
+    if (_tokensToMintFromERC20s != 0)
+      IERC20(_v1Token).transferFrom(msg.sender, address(this), _tokensToMintFromERC20s);
+
+    // Transfer v1 unclaimed tokens to this contract from the msg sender if needed.
+    if (_tokensToMintFromUnclaimedBalance != 0)
+      v1TicketBooth.transfer(
+        msg.sender,
+        projectId,
```

Comment:
Isn't this assuming project id v1 == id v3, which isn't true in real setting (imo, test should use different project id, iirc, that's how we did for the (deprecated) migration terminal)